### ISSUE TITLE
User-editable Tile Stamp storage

### ISSUE THREAD

**Issue by eishiya** [mapeditor/tiled#3273]
Currently, Tile Stamps are stored in a custom JSON format that cannot be opened for editing in Tiled. To modify a stamp even slightly, you have to stamp it on a map, make your changes, create a new stamp from this, and then delete the old stamp. This can get quite annoying. If you want to make versions of stamps using a different tileset, you have to stamp all of them on my map one by one, replace the tileset, and then make new stamps one by one.

I propose using Tiled maps (JSON/TMJ or TMX) to store stamps instead. Each map would contain a stamp, including all of its variants. To distinguish between different variants, each should be stored in a top-level group. Any groups within these groups would just be ignored, the inner structure of the stamp can be treated as just a flat arrangement of layers, as is currently done for Automapping layers. The probability of each variant would be stored in a custom float `probability` property on the group. The base stamp's name can be stored as a `name` custom property on the map (filenames can't store every possible name, which is presumably why the current format has a `name` property).

Here's an example of what such a map might look like:
![image](https://user-images.githubusercontent.com/70545646/153903622-675cfc2d-af22-400f-beda-39b56ac305df.png)

Since each variant is a separate group, there is no ambiguity with determining which variant a given tile belongs to, so Stamps consisting of disjoint tiles would work just fine with this. When the stamps are loaded, they should be trimmed, so that the empty space around them in the map doesn't affect their position. When the stamp files are written, the stamps should be arranged in rows according to their bounding box, with 1-2 tiles of space between them, so that they don't overlap each other. This makes for a much more pleasant editing experience. Another option is to size the map to the size of the largest stamp and use layer offsets on the groups to position the stamps, which would reduce the amount of empty cells in the layers, but I think this would be more confusing for users not used to working with offset layers.

Additional features that this format could enable without any extra UI for the Tile Stamps panel:
- Naming individual variants (edit the group names)
- Bulk editing of stamp probabilities (select multiple groups at once, edit the probability property)
- Reordering variants (reorder the groups)

If Tiled ever makes it easier to copy entire layers or groups of layers from one map to another, or if a user has a script to aid with this, this system would also make it easier to move variants between different stamps or to convert a variant into its own stamp, which could solve #2982.

In addition, this format could enable setting per-stamp anchors as a pair of custom properties on the group, or with a special layer within the group. This could be even more flexible than the Stamp Brush option suggested in #3250, as the anchor could be placed anywhere within the stamp, rather than being confined to one of nine positions relative to the centre of the stamp. If done with a special layer, then this could be a layer with a custom bool property `anchor` set to `true`, and the first found non-empty tile within this layer would serve as the anchor position. I think this would be the most intuitive method, and it would keep the anchor in place even if the whole stamp is moved within the map. Using a pair of custom properties to store the anchor would require deciding whether the point is a map coordinate or a coordinate relative to the stamp's bounding box, and would need to be manually updated if the user moves the stamp.

**Comment by UliAbo** [comment]
I support the urgency of this issue. For our situations it currently feels like impossible to work with stamps across several users or am I wrong? Even when setting the same stamp folder and copying the stamps there, they don't get loaded. 

If I understand @eishiya 's points above correctly, here in this thread they ask for more than a "simple" storage of stamps? Would this maybe a first lean step? 

So in other words: It would be very helpful for us if stamps get saved correctly across different users. Otherwise we can't work on a team on a map using the same stamps. I'm happy if I'm proven wrong or maybe you guys know some workarounds :)

**Comment by eishiya** [comment]
This issue has nothing to do with sharing stamps, and would probably not change anything related to that, it would only change the format in which stamps are saved.

You can already share your stamps along with your project by setting your stamps folder to be within your project, and storing the stamps there instead of the default global folder.

**Comment by UliAbo** [comment]
Thanks @eishiya . You are right, it works now for us using projects. I was mislead there.
